txpipe/noise_maps.py

```diff
@@ -541,3 +541,291 @@ def finalize_mappers(self, pixel_scheme, mappers):
+
+            # Compute which pixel each object is in
+            ra = data['ra']
+            dec = data['dec']
+            orig_pixels = device_put(pixel_scheme.ang2pix(ra, dec))
+            pixels = device_put(index_map[orig_pixels])
+
+            # This is how you do RNG with JAX. We use subkey for this RNG operation
+            # and then key is passed forward for the next operation
+            key, subkey = random.split(key)
+
+            # randomly select a half for each lens bin object
+            # random.bernoulli returns True/False arrays. Convert that
+            # to an integer array (both on the GPU) by multiplying by 1
+            split = 1 * random.bernoulli(subkey, 0.5, (n, clustering_realizations))
+
+            # random rotations of the g1, g2 values
+            key, subkey = random.split(key)
+            phi = random.uniform(subkey, shape=(lensing_realizations, n), minval=0, maxval=2*jnp.pi)
+            cos = jnp.cos(phi)
+            sin = jnp.sin(phi)
+            g1r = jnp.transpose(cos * g1 + sin * g2)
+            g2r = jnp.transpose(-sin * g1 + cos * g2)
+
+            # masks showing which pixels to fill in
+            pix_mask = pixels >= 0
+            sb_mask = (source_bin >= 0) & pix_mask
+            
+
+            # jax.jit doesn't like masks inside masks so we have to calculate these in
+            # advance instead of doing that within the jitted functions
+            masked_pixels = pixels[sb_mask]
+            masked_source_bin = source_bin[sb_mask]
+            masked_g1r = g1r[sb_mask]
+            masked_g2r = g2r[sb_mask]
+            masked_weights = weights[sb_mask]
+            lb_mask = sb_mask & (lens_bin >= 0)
+            pixels_lb_mask = pixels[lb_mask]
+            lens_bin_lb_mask = lens_bin[lb_mask]
+            split_lb_mask = split[lb_mask]
+
+            # Accumulate into the total noise maps. Under JAX this can't be an in-place
+            # operation, so we have to replace G1 each time. Under the hood this may
+            # be happening in-place, I think it depends.
+            G1 = GN_add_jit(G1, masked_pixels, masked_source_bin, masked_g1r)
+            G2 = GN_add_jit(G2, masked_pixels, masked_source_bin, masked_g2r)
+            GW = GW_add_jit(GW, masked_pixels, masked_source_bin, masked_weights)
+            ngal_split = ngal_split_add_jit(ngal_split, pixels_lb_mask, clustering_realizations, split_lb_mask, lens_bin_lb_mask)
+            # TODO: Currently breaks with clustering_realizations > 1
+
+        # Now we have finished looping through the data, we sum everything over the
+        # different processes to the root process
+        if self.comm is not None:
+            import mpi4jax
+            from mpi4py import MPI
+            mpi4jax.reduce(G1, MPI.SUM, root=0)
+            mpi4jax.reduce(G2, MPI.SUM, root=0)
+            mpi4jax.reduce(GW, MPI.SUM, root=0)
+            mpi4jax.reduce(ngal_split, MPI.SUM, root=0)
```

Comment:
Hi @joezuntz , chiming in very randomly (I'm the author of `mpi4jax` and was recently looking at who uses the library for pure curiosity)

I noticed those few lines which are probably not what you want.

`mpi4jax` follows jax principles of being purely functional and not modifying in-place inputs. So those few lines are scheduling the reduction operation and they are  discarding the result.

I think you'd want to do something more like
```python
G1, token = mpi4jax.reduce(G1, MPI.SUM, root=0)
G2, token = mpi4jax.reduce(G2, MPI.SUM, root=0, token=token)
GW, token = mpi4jax.reduce(GW, MPI.SUM, root=0, token=token)
ngal_split, token = mpi4jax.reduce(ngal_split, MPI.SUM, root=0, token=token)
```

(the tokens are a way to enforce those operations to execute in a well-defined order, because jax does not necessarily execute operations in the same order you write them. This sometimes leads to MPI deadlocks if you do not use tokens. However, It might be that in this very simple case that's not needed.)

Reply:
@PhilipVinc thanks so much, this is v useful!  Very kind of you to take the time to let us know this!

@Lhior could you open a new pull request to make these changes?

Reply:
no probs.
By the way, just a suggestion: the block of logic above (the body of the loop for `for (s, e, data) in it:`) looks like could be written in a single `jax.jit` kernel.

Every jax operation that is not fitted has a fairly high overhead, so unless you have large > 10^6 elements you most likely are paying a considerable price. When jitting you pay the overhead only once.